### ISSUE TITLE
Dataclass `__init__` is not using descriptor `__set__` overloads to determine types

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#3866]
**Describe the bug**
Given the following code:

```python
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Protocol, TypeVar, overload
from typing_extensions import reveal_type

_T = TypeVar('_T')


class Helper:
    ...


class Mapped(Protocol[_T]):
    def join(self, other: Any) -> Any:
        ...

    @overload
    def __get__(self, instance: None, owner: Any) -> Mapped[_T]:
        ...

    @overload
    def __get__(self, instance: object, owner: Any) -> _T:
        ...

    @overload
    def __set__(self, instance: None, value: Helper) -> None:
        ...

    @overload
    def __set__(self, instance: object, value: _T) -> None:
        ...


@dataclass
class MyThing:
    id: Mapped[int]


def main(my_thing: MyThing) -> None:
    reveal_type(MyThing.id)  # Mapped[int]
    reveal_type(my_thing.id)  # int

    helper = Helper()
    MyThing.id = 0  # no error
    reveal_type(MyThing.id)  # Literal[0]
    MyThing.id = helper
    reveal_type(MyThing.id)  # Helper
    MyThing(id=0)  # error: Literal[0] is incompatible with protocol Mapped[int]
```

1. Line 45 does not produce an error
2. Line 49 produces an error

**Expected behavior**
1. Line 45 should produce an error
2. Line 49 should not produce an error

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command line

Which version?
1.1.268

**Additional context**
If one of the overloads is commented out, one of the two problematic lines behaves as expected:

1. Commenting out the overload at line 26 fixes the error at line 49
2. Commenting out the overload at line 30 causes an error to be produced at line 45

**Comment by erictraut** [comment]
The use of descriptors within decorators is problematic. The authors of PEP 557 admitted that they hadn't considered this combination when designing data classes, and there are bunch of unspecified and unexpected behaviors that occur when the two are used together.

We had multiple [long discussions](https://github.com/microsoft/pyright/discussions/2958) on this already in context of [PEP 681](https://peps.python.org/pep-0681/) (Data Class Transforms). These discussions resulted in consensus around a set of behaviors that I subsequently implemented in pyright. The behaviors you're seeing in the sample above are a result of this.

The assumption is that if you want to use descriptor objects in a dataclass, those should be assigned as class variables. In your sample above, it's not clear to me who is intended to instantiate a `Mapped` descriptor object and assign it to the class variable `id`.

**Comment by debonte** [comment]
FYI, there's a more concise discussion about defining/standardizing the behavior of descriptor-typed fields here: https://github.com/python/cpython/issues/91330

In particular, we called out that "if a field is annotated with a descriptor type but is *not* assigned a descriptor object as its default value, it acts like a non-descriptor field," which is what you are seeing.

**Comment by bryanforbes** [comment]
I'm still not sure the behavior is correct (and perhaps I'm misunderstanding/misinterpreting what's been said in these other threads). Taking my code from above and assigning a default descriptor value:

```python
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Generic, TypeVar, overload
from typing_extensions import reveal_type

_T = TypeVar('_T')


class Helper:
    ...


class Mapped(Generic[_T]):
    def __init__(self, *, init: bool = True) -> None:
        ...

    def join(self, other: Any) -> Any:
        ...

    @overload
    def __get__(self, instance: None, owner: Any) -> Mapped[_T]:
        ...

    @overload
    def __get__(self, instance: object, owner: Any) -> _T:
        ...

    def __get__(self, instance: Any, owner: Any) -> Any:
        ...

    @overload
    def __set__(self, instance: None, value: Helper) -> None:
        ...

    @overload
    def __set__(self, instance: object, value: _T) -> None:
        ...

    def __set__(self, instance: Any, value: Any) -> None:
        ...


@dataclass
class MyThing:
    id: Mapped[int] = Mapped[int]()


def main(my_thing: MyThing) -> None:
    reveal_type(MyThing.id)  # Mapped[int]
    reveal_type(my_thing.id)  # int

    helper = Helper()
    MyThing.id = 0  # no error
    reveal_type(MyThing.id)  # Literal[0]
    MyThing.id = helper
    reveal_type(MyThing.id)  # Helper
    MyThing(id=0)  # error: Literal[0] is incompatible with protocol Mapped[int]
```

Line 54 still does not produce an error and line 58 still produces an error.

**Comment by debonte** [comment]
What is the purpose of the `None` `__set__` overload that takes a `Helper` instance? Are you wanting to overwrite the descriptor instance? That seems strange to me.

**Comment by bryanforbes** [comment]
_sigh_ It must be time for me to be done for the day. For some reason, I got it in my head that `__set__` worked at the class level like `__get__`. Sorry for the noise.

**Comment by debonte** [comment]
No problem. I'm not sure why that confused Pyright, but I'm glad you're unblocked.

Btw, speaking of the `None` overload for `__get__`:

```python
@overload
def __get__(self, instance: None, owner: Any) -> Mapped[_T]:
    ...
```

One of the descriptor-typed field behaviors that was recently documented is that the value returned by `__get__` when `instance` is `None` will be used as the default value of fields of this descriptor type. This was already the behavior at runtime -- we simply documented it.

So you may want to consider returning `_T` here rather than `Mapped[_T]`.
